### PR TITLE
New Animation API

### PR DESCRIPTION
**Pull request by lucasdinonolte**
## TL;DR

Adds a draw loop helper to mechanic core. This helper is exposed to design functions through the engine. The goal of this is to remove boilerplate from design functions.

Implements the changes explored and discussed in #156 

### Core
- adds `frameRate` setting to mechanic instance (defaults to 60)
  - frameRate is passed to the webm writer and used to prepare a draw loop helper
- adds `defaultSettings` file that is merged with user provided settings (mostly to create visibility for the default settings)
- adds a singleton drawloop helper to mechanic core
  - whenever a new instance of mechanic is created the draw loop helper is configured with the frameRate setting
  - a singleton was chosen to be entirely sure we can reset the draw loop before reloading a function
- adds `registerFrameCallback` and `registerDoneCallback` methods to mechanic core
  - these can be used by engines to register their onFrame and onDone methods
  - this is needed so common logic (like resetting the draw loop on Done) can be handled by core instead of being duplicated in every engine
  - imho this also streamlines the readbility of each engine a tiny bit
- adds `callbacksForEngine` to mechanic core
  - this returns an object of callbacks (`frame`, `done`,  `drawLoop` …) that an engine can pass to a function
  - the function takes an overwrite-argument to allow engines to overwrite callbacks if they need 
  - engine p5 uses this to output an error if you try to use mechanic's draw loop

### All engines
- Refactor engines to work with the updated API (`registerFrameCallback`, `registerDoneCallback` and `callbacksForEngine`)
- Refactor engines to expose `frame`, `done` and `drawLoop` to the design function directly, instead of on the `mechanic` object passed in
  - `state` and `setState` remain on `mechanic` to avoid namespace collisions with React
  - for consistency this is true for all engines

### Engine Canvas
- Add `getCanvas` helper to engine that provides the user with a density-aware preview canvas
  - this is just for preview, this does not implement different pixel densities on export to keep the scope of this PR limited
  - pixel Density for export should be implemented on another PR (#144 did some exploration)
- When using `getCanvas` a user no longer needs to pass `canvas` to `done` or `frame`
  - `done` and `frame` still accept a canvas to be passed to them, to keep it backwards compatible with functions creating their own canvas

### Engine p5
- Call `sketch.frameRate` with `mechanic.settings.frameRate` before starting rendering on the sketch

### DSI Logo Maker
- Refactor all functions to new API
  - `Logo Animated SVG` renders a bit stuttery, but this does not seem to be related to the changes. It's the same on `main`
  
### Create-Mechanic
- Refactor all function-templates and examples to new API

## Discussion

I tried to keep this PR as concise as possible, but given it needs to touch almost all parts of mechanic it has become fairly large. I also tried to be expressive with code comments and added `TODO` labels where I feel unsure about a decision I've made. Looking forward to discussing this.

### CODE REVIEW

**Review by fdoflorenzano — COMMENTED** [review]
This is an initial set of comments. For now I just checked the core and engine changes, I'll leave the checking of the actual function updates for a bit later.

File: packages/core/src/mechanic-drawloop.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  callback;
+
+const epsilon = 5;
+
+class MechanicDrawloop {
+  /**
+   * Drawloop constructor. The drawloop is realized as a singleton. Making sure
+   * there is only one instance of the drawloop is needed so we can easily clear
+   * any pending animation frames when it's time to update a design function due
+   * to code changes or new input values.
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic-drawloop.js:17]: Yes, Mechanic Core adds a `this.drawLoop.stop()`. I mean technically you shouldn't be able to get a second instance of drawloop, since we export and instance and not the class.

File: packages/engine-canvas/index.js [hunk 2]
```diff
@@ -7,40 +8,66 @@ export const run = (functionName, func, values, config) => {
...
+
   const mechanic = new Mechanic(func.settings, values, config);
 
-  let isElAdded = false;
-  const onFrame = (el) => {
+  // This ensures backwards compatibility for function's building their own
+  // canvas
+  //
+  // TODO: Do we want to keep it backwards or compatible or do we just remove
+  // the old style, given that the new animation API is breaking anyways?
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/engine-canvas/index.js:20]: Very good point! Also keeping the ability to bring your own canvas would potentially allow users to use the canvas engine with WebGL if they need too (and until we decide if we want to add a specialized engine for 3D canvases).

File: packages/core/src/mechanic-drawloop.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
+let instance,
+  isPlaying = false,
+  frameRate,
+  fpsInterval,
+  frameCount,
+  lastFrameTime,
+  raf,
+  callback;
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic-drawloop.js:8]: I guess it's even unnecessary to store them here. As we make sure there can only ever be one instance of drawLoop in the constructor we can move them into the class.

File: packages/core/src/mechanic-drawloop.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
+let instance,
+  isPlaying = false,
+  frameRate,
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic-drawloop.js:3]: ✅

File: packages/core/src/mechanic-drawloop.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    isPlaying = false;
+    if (raf) window.cancelAnimationFrame(this.raf);
+  }
+
+  /**
+   * Renders a single frame of the drawloop
+   * @param {number} frame - The number of the frame to render
+   */
+  renderFrame(frame) {
+    if (callback) callback.call(null, frame);
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic-drawloop.js:73]: you're right. removing it.

File: packages/engine-canvas/index.js [hunk 6]
```diff
@@ -7,40 +8,66 @@ export const run = (functionName, func, values, config) => {
...
+  const checkForCanvas = el => {
+    if (!el && !preparedCanvas) {
+      throw new Error(
+        `You need to call getCanvas() before calling frame() or pass your own canvas element as an argument to frame().`
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/engine-canvas/index.js:24]: ✅

File: packages/core/src/mechanic.js [hunk 7]
```diff
@@ -86,14 +102,73 @@ export class Mechanic {
...
+   * Convenience function to collect all callbacks that an engine can pass to a
+   * design function. An engine can pass overwrites to inject custom behavior
+   * into the callbacks. This is currently used by engine-p5 to disable using
+   * mechanic.drawLoop within engine-p5.
+   *
+   * @param {object} overwrites - An object with overwrites for the callbacks
+   *
+   * @TODO: Could this be merged with the function above?
+   */
+  callbacksForEngine(overwrites = {}) {
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic.js:115]: agreed

File: packages/core/src/mechanic.js [hunk 8]
```diff
@@ -86,14 +102,73 @@ export class Mechanic {
...
 
+  /**
+   * Convenience function to collect all callbacks that an engine can pass to a
+   * design function. An engine can pass overwrites to inject custom behavior
+   * into the callbacks. This is currently used by engine-p5 to disable using
+   * mechanic.drawLoop within engine-p5.
+   *
+   * @param {object} overwrites - An object with overwrites for the callbacks
+   *
+   * @TODO: Could this be merged with the function above?
```
[the remark this answers is not in the document]
Comment (lnolte) [reply packages/core/src/mechanic.js:113]: ✅

**Comment by lnolte** [comment]
@fdoflorenzano thanks for the initial set of comments. I just addressed them. From my point of view this should now be ready for you to play with writing some design functions using the new API to see how it feels :blush:

/cc @runemadsen

**Review by fdoflorenzano — COMMENTED** [review]
Thank you @lnolte ! Most comments are address, so yeah I also think it's a good point to check the actual functions. I'll try and do that tomorrow :)

File: packages/engine-canvas/prepare-canvas.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
+export const buildCanvasDimensions = ({
```
Comment (fdoflorenzano) [inline packages/engine-canvas/prepare-canvas.js:1]: This file needs to get added to the engine's package.json!

**Review by fdoflorenzano — CHANGES_REQUESTED** [review]
Found something! (Will continue reviewing!)

**Review by fdoflorenzano — COMMENTED** [review]
Checked all functions finally!

**Review by fdoflorenzano — APPROVED** [review]
Let's wait for #169 into this first and then merge if @runemadsen also approves!

**Review by runemadsen — APPROVED** [review]
Yezzzz, let's get this sucker in!
